Termination Measures for Loops and Functions

JM4ier opened the pull request.

vakaras commented on prusti-viper/src/encoder/mir/procedures/encoder/loops.rs:
```diff
@@ -100,4 +100,81 @@ impl<'p, 'v: 'p, 'tcx: 'v> super::ProcedureEncoder<'p, 'v, 'tcx> {
         )?;
         Ok(statement)
     }
+    pub(super) fn encode_loop_variant(
```
As discussed this probably should be part of the `LoopInvariant` desugaring.

vakaras commented on prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:
```diff
@@ -1801,6 +1925,21 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         }
 
+        // Encode loop variants.
+        let mut loop_variant_blocks = self.specification_blocks.loop_variant_blocks().clone();
```
This should be done together with the loop invariant.

vakaras commented on prusti-viper/src/encoder/mir/procedures/encoder/specification_blocks.rs:
```diff
@@ -20,6 +20,7 @@ pub(super) struct SpecificationBlocks {
     ///
     /// FIXME: Add a check that ensure that the blocks are one after another.
     loop_invariant_blocks: BTreeMap<mir::BasicBlock, LoopInvariantBlocks>,
+    loop_variant_blocks: BTreeMap<mir::BasicBlock, LoopInvariantBlocks>,
```
Couldn't these be merged?

vakaras reviewed (changes requested):
Overall looks fine to me. Major things to fix:

* [ ] Rebase/merge master.
* [ ] Merge `LoopVariant` into `LoopInvariant`.

Aurel300 commented on prusti-tests/tests/verify/pass/terminating_fib.rs:
```diff
@@ -0,0 +1,16 @@
+
+fn main() {}
+
+#[pure(Int::new(x))]
```
What is this? Is this the termination measure? Why is it attached to the `pure` attribute?

JM4ier replied:
yes, it is a termination measure.
The PR makes it a requirement for recursive pure functions, or recursive functions with the `#[terminates]` attribute.
I also specifically added it to the `#[pure]` attribute because nonterminating pure expressions in pre/post conditions can cause unsoundness

Aurel300 replied:
Still I don't think the UX is good. It should be clear when reading the contract of a method at least roughly what it means. So, if a termination measure is required for a pure function and there is none provided, we should tell the user to write a `#[terminates(...)]` attribute.

JM4ier committed:
Rebased Termination Code onto master

features:
- functions can be annotated with `#[terminates]`
- pure functions need to terminate
- loop variants with the `body_variant!(...)` macro
- ghost! blocks need to terminate

vakaras commented on prusti-contracts/prusti-contracts-impl/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+#![no_std]
```
This file was deleted on `master`.

vakaras commented on prusti-contracts/prusti-contracts-internal/src/lib.rs:
```diff
@@ -0,0 +1,107 @@
+use proc_macro::TokenStream;
+use prusti_specs::{rewrite_prusti_attributes, SpecAttributeKind};
```
This file also does not exist on `master`.

vakaras commented on prusti-interface/src/environment/mod.rs:
```diff
@@ -134,6 +135,60 @@ impl<'tcx> Environment<'tcx> {
             .unwrap()
     }
 
+    pub fn is_recursive(&self, proc_def_id: ProcedureDefId) -> bool {
```
Isn't this function dead code? If yes, please delete it.

vakaras commented on prusti-viper/src/encoder/mir/type_layouts/interface.rs:
```diff
@@ -11,14 +11,23 @@ pub(crate) trait MirTypeLayoutsEncoderInterface<'tcx> {
         &self,
         ty: ty::Ty<'tcx>,
     ) -> SpannedEncodingResult<vir_high::Expression>;
+    fn encode_type_size_expression_with_reps_vir(
```
Dead code?

vakaras reviewed (changes requested):
Looks good module the comments. Thanks!

vakaras commented on prusti-contracts/prusti-specs/src/rewriter.rs:
```diff
@@ -110,12 +112,17 @@ impl AstRewriter {
         //   terminator in MIR has a span set to the one character just after
         //   the identifier
+        let (return_type, return_modifier) = if spec_type == SpecItemType::Termination {
+            (quote! {Int}, quote! {Int::new(0) + })
```
Could you try using `quote_spanned!` here to see whether that fixes the `test_failing_crate` test?

JM4ier committed:
termination fixup

Commit:
improved termination error messages

Commit:
ghost code usage example employing a lemma function that is not needed during runtime

JM4ier committed:
clippy fix

vakaras committed:
Fix UI test.

vakaras committed:
Introduce versioning for cache.

vakaras committed:
Rustfmt.

vakaras committed:
Implement terminates(trusted).

vakaras committed:
Fix UI tests.

vakaras committed:
Implement snapshot equality.

vakaras committed:
Fix test annotations.

vakaras committed:
Debug non-termination.

vakaras committed:
Fix bors.
